core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java

```diff
@@ -429,22 +405,61 @@ public ClientFactoryBuilder useHttp1Pipelining(boolean useHttp1Pipelining) {
      */
     public ClientFactoryBuilder connectionPoolListener(
             ConnectionPoolListener connectionPoolListener) {
-        this.connectionPoolListener = requireNonNull(connectionPoolListener, "connectionPoolListener");
+        option(ClientFactoryOption.CONNECTION_POOL_LISTENER,
+               requireNonNull(connectionPoolListener, "connectionPoolListener"));
         return this;
     }
 
     /**
      * Sets the {@link MeterRegistry} which collects various stats.
      */
     public ClientFactoryBuilder meterRegistry(MeterRegistry meterRegistry) {
-        this.meterRegistry = requireNonNull(meterRegistry, "meterRegistry");
+        option(ClientFactoryOption.METER_REGISTRY, requireNonNull(meterRegistry, "meterRegistry"));
         return this;
     }
 
     /**
-     * Returns a newly-created {@link ClientFactory} based on the properties of this builder.
+     * TBW.
      */
-    public ClientFactory build() {
+    public <T> ClientFactoryBuilder option(ClientFactoryOption<T> option, T value) {
+        requireNonNull(option, "option");
+        requireNonNull(value, "value");
+        return option(option.newValue(value));
+    }
+
+    /**
+     * TBW.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> ClientFactoryBuilder option(ClientFactoryOptionValue<T> optionValue) {
+        requireNonNull(optionValue, "optionValue");
+        final ClientFactoryOption<?> opt = optionValue.option();
+        if (opt == ClientFactoryOption.EVENT_LOOP_SCHEDULER) {
+            eventLoopSchedulerFactory(eventLoopGroup -> (EventLoopScheduler) optionValue.value());
+        } else if (opt == ClientFactoryOption.ADDRESS_RESOLVER_GROUP) {
+            addressResolverGroupFactory(
+                    eventLoopGroup -> (AddressResolverGroup<InetSocketAddress>) optionValue.value());
```

Comment:
Why don't we just replace `EVENT_LOOP_SCHEDULER` and `ADDRESS_RESOLVER_GROUP` with `EVENT_LOOP_SCHEDULER_FACTORY` and `ADDRESS_RESOLVER_GROUP_FACTORY`?